File: python/src/sello/service.py
from __future__ import annotations

import inspect
import json
import os
from concurrent.futures import Future, ThreadPoolExecutor, wait
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable, Optional, Union
from urllib.request import urlopen

from . import logs
from .keys import normalize_ed25519_public_key, normalize_service_key
from .receipt import BuiltReceipt, build_receipt, canonical_json_bytes
from .token import verify_sello_jws_token


class SelloDeniedError(Exception):
    def __init__(self, receipt: BuiltReceipt):
        super().__init__("Sello request denied")
        self.receipt = receipt


@dataclass
class _ResolvedConfig:
    service: str
    service_kid: bytes
    service_private_key: bytes
    token_issuer_public_key: bytes
    log: Union[logs.MemoryLog, logs.HttpLog]
    fallback_sello_logs: list[str]
    submit_mode: str
    now: Callable[[], str]
    on_receipt: Optional[Callable[[dict[str, Any]], None]]
    on_submit_error: Optional[Callable[[BaseException], None]]
    on_drop: Optional[Callable[[dict[str, Any]], None]]
    max_pending: int
    concurrency: int


class SelloReceipts:
    def __init__(self, config: _ResolvedConfig):
        self._config = config
        self._executor = ThreadPoolExecutor(max_workers=config.concurrency)
        self._futures: list[Future[Any]] = []

    def tool(self, action_type: str, handler: Optional[Callable[..., Any]] = None, **options: Any):
        if not action_type:
            raise ValueError("Sello action type must be a non-empty string")

        def decorate(fn: Callable[..., Any]):
            if inspect.iscoroutinefunction(fn):
                async def async_wrapped(request: Any):
                    return await self._run_async(action_type, fn, request, options)
                return async_wrapped

            def wrapped(request: Any):
                return self._run_sync(action_type, fn, request, options)
            return wrapped

        return decorate(handler) if handler is not None else decorate

    def flush(self) -> None:
        futures = list(self._futures)
        self._futures.clear()
        if futures:
            wait(futures)

    async def _run_async(self, action_type: str, handler: Callable[[Any], Any], request: Any, options: dict[str, Any]):
        base = self._base(action_type, request, options)
        if await _maybe_await(options.get("is_denied", lambda _: False)(request)):
            response = await _maybe_await(options["denied_response"](request)) if "denied_response" in options else None
            receipt = self._emit(base, b"", "denied")
            if "denied_response" not in options:
                raise SelloDeniedError(receipt)
            return response
        try:
            response = await handler(request)
        except Exception as error:
            self._emit(base, _error_bytes(error, options), "error", error=error)
            raise
        self._emit(base, _output_bytes(response, options), "success", response=response)
        return response

    def _run_sync(self, action_type: str, handler: Callable[[Any], Any], request: Any, options: dict[str, Any]):
        base = self._base(action_type, request, options)
        if options.get("is_denied", lambda _: False)(request):
            response = options["denied_response"](request) if "denied_response" in options else None
            receipt = self._emit(base, b"", "denied")
            if "denied_response" not in options:
                raise SelloDeniedError(receipt)
            return response
        try:
            response = handler(request)
        except Exception as error:
            self._emit(base, _error_bytes(error, options), "error", error=error)
            raise
        self._emit(base, _output_bytes(response, options), "success", response=response)
        return response

    def _base(self, action_type: str, request: Any, options: dict[str, Any]) -> dict[str, Any]:
        token = _resolve_authorization_token(request, options)
        verified = verify_sello_jws_token(token, self._config.token_issuer_public_key)
        sello_logs = verified.sello_logs or self._config.fallback_sello_logs
        if self._config.log.log_url not in sello_logs:
            raise ValueError("Sello log must be listed in the token's sello_logs")
        return {
            "authorization_token_bytes": verified.authorization_token_bytes,
            "owner_hpke_public_key": verified.owner_hpke_public_key,
            "sello_logs": sello_logs,
            "action_type": action_type,
            "action_input_bytes": _input_bytes(request, options),
            "timestamp": self._config.now(),
        }

    def _emit(self, base: dict[str, Any], action_output_bytes: bytes, result_status: str, response: Any = None, error: Optional[BaseException] = None) -> BuiltReceipt:
        receipt = build_receipt(
            **base,
            action_output_bytes=action_output_bytes,
            result_status=result_status,
            service_kid=self._config.service_kid,
            service_private_key=self._config.service_private_key,
            service_identifier=self._config.service,
            log_url=self._config.log.log_url,
        )
        if self._config.on_receipt:
            event = {"result_status": result_status, "receipt": receipt}
            if response is not None:
                event["response"] = response
            if error is not None:
                event["error"] = error
            self._config.on_receipt(event)
        self._submit(receipt.envelope, base["timestamp"])
        return receipt

    def _submit(self, envelope: bytes, integrated_time: str) -> None:
        if self._config.submit_mode == "await":
            self._config.log.append(envelope, integrated_time)
            return
        self._futures = [future for future in self._futures if not future.done()]
        if len(self._futures) >= self._config.max_pending:
            if self._config.on_drop:
                self._config.on_drop({"envelope": envelope, "integrated_time": integrated_time, "reason": "queue_full"})
            return
        future = self._executor.submit(self._config.log.append, envelope, integrated_time)
        if self._config.on_submit_error:
            future.add_done_callback(lambda done: done.exception() and self._config.on_submit_error(done.exception()))
        self._futures.append(future)


def service(config: Optional[Union[str, dict[str, Any]]] = None, **kwargs: Any) -> SelloReceipts:
    values: dict[str, Any] = {}
    if isinstance(config, str):
        values["service"] = config
    elif isinstance(config, dict):
        values.update(config)
    elif config is not None:
        raise ValueError("service config must be a service id string or dict")
    values.update(kwargs)
    return SelloReceipts(_resolve_config(values, os.environ))


def _resolve_config(values: dict[str, Any], env: os._Environ[str]) -> _ResolvedConfig:
    service_id = values.get("service") or env.get("SELLO_SERVICE_ID")
    if not service_id:
        raise ValueError("Sello setup missing SELLO_SERVICE_ID")
    service_kid, service_private_key = normalize_service_key(
        values.get("service_key") or env.get("SELLO_SERVICE_KEY"),
        values.get("service_kid") or env.get("SELLO_SERVICE_KID"),
    )
    log = values.get("log")
    if log is None:
        log_url = env.get("SELLO_LOG_URL")
        if not log_url:
            raise ValueError("Sello setup missing SELLO_LOG_URL")
        log = logs.http(log_url, endpoint=env.get("SELLO_LOG_ENDPOINT"))
    issuer = values.get("token_issuer") or values.get("token_issuer_public_key") or env.get("SELLO_TOKEN_ISSUER_PUBLIC_KEY")
    if issuer is None:
        jwks = values.get("token_issuer_jwks") or env.get("SELLO_TOKEN_ISSUER_JWKS")
        if not jwks:
            raise ValueError("Sello setup missing token issuer")
        issuer = _fetch_jwks_ed25519_public_key(jwks)
    submit = values.get("submit") or env.get("SELLO_SUBMIT_MODE") or "background"
    submit_mode = submit.get("mode", "background") if isinstance(submit, dict) else submit
    return _ResolvedConfig(
        service=str(service_id),
        service_kid=service_kid,
        service_private_key=service_private_key,
        token_issuer_public_key=normalize_ed25519_public_key(issuer, "token_issuer"),
        log=log,
        fallback_sello_logs=values.get("fallback_sello_logs") or [log.log_url],
        submit_mode=submit_mode,
        now=values.get("now") or _now_utc_seconds,
        on_receipt=values.get("on_receipt"),
        on_submit_error=values.get("on_submit_error"),
        on_drop=values.get("on_drop"),
        max_pending=(submit.get("max_pending", 1000) if isinstance(submit, dict) else 1000),
        concurrency=(submit.get("concurrency", 4) if isinstance(submit, dict) else 4),
    )


def _resolve_authorization_token(request: Any, options: dict[str, Any]) -> Union[str, bytes]:
    source = options.get("authorization_token")
    if callable(source):
        return source(request)
    if source is not None:
        return source
    if isinstance(request, dict):
        token = request.get("authorizationToken") or request.get("authorization")
        if token:
            return _strip_bearer(token)
        headers = request.get("headers")
        if isinstance(headers, dict):
            token = headers.get("authorization") or headers.get("Authorization")
            if token:
                return _strip_bearer(token)
    raise ValueError("Sello authorization token not found")


def _strip_bearer(value: Union[str, bytes]) -> Union[str, bytes]:
    if isinstance(value, str) and value.startswith("Bearer "):
        return value[len("Bearer "):]
    return value


def _input_bytes(request: Any, options: dict[str, Any]) -> bytes:
    canonicalizer = options.get("canonicalize_input")
    return canonicalizer(request) if canonicalizer else canonical_json_bytes(request)


def _output_bytes(response: Any, options: dict[str, Any]) -> bytes:
    canonicalizer = options.get("canonicalize_output")
    return canonicalizer(response) if canonicalizer else canonical_json_bytes(response)


def _error_bytes(error: BaseException, options: dict[str, Any]) -> bytes:
    canonicalizer = options.get("canonicalize_error")
    return canonicalizer(error) if canonicalizer else canonical_json_bytes({"name": error.__class__.__name__, "message": str(error)})


def _now_utc_seconds() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


async def _maybe_await(value: Any) -> Any:
    if inspect.isawaitable(value):
        return await value
    return value


def _fetch_jwks_ed25519_public_key(jwks_url: str) -> bytes:
    with urlopen(jwks_url) as response:
        jwks = json.loads(response.read().decode("utf-8"))
    for key in jwks.get("keys", []):
        if key.get("kty") == "OKP" and key.get("crv") == "Ed25519" and isinstance(key.get("x"), str):
            from ._crypto import base64url_decode
            return base64url_decode(key["x"], "JWKS x")
    raise ValueError("token issuer JWKS must contain an Ed25519 OKP key")

File: sdks/typescript/src/sdk/service.ts
import { type BuiltReceipt, buildReceipt } from "../service/create-receipt.ts";
import { verifySelloJwsToken } from "../token/jws-profile.ts";
import { type CanonicalLogUrl, assertCanonicalLogUrl, logUrlsEqual } from "../log/canonical-url.ts";
import { type ResultStatus } from "../receipt/body.ts";
import { canonicalJsonBytes } from "../mcp/middleware.ts";
import {
  type KeyInput,
  type ServiceKeyInput,
  decodeBase64url,
  normalizeEd25519PublicKey,
  normalizeServiceKey,
} from "./keys.ts";
import { type SdkSubmissionLog, http } from "./logs.ts";
import {
  BackgroundReceiptPublisher,
  type DropEvent,
  type SelloSubmitOptions,
  type SubmitMode,
} from "./publisher.ts";

export type SelloToolHandler<Request, Response> = (
  request: Request,
) => Response | Promise<Response>;

export type SelloValueOrGetter<Request, Value> =
  | Value
  | ((request: Request) => Value);

export type SelloReceiptEvent<Response = unknown> = {
  resultStatus: ResultStatus;
  receipt: BuiltReceipt;
  response?: Response;
  error?: unknown;
};

export type SelloToolOptions<Request, Response> = {
  authorizationToken?: SelloValueOrGetter<Request, string | Uint8Array>;
  canonicalizeInput?: (request: Request) => Uint8Array;
  canonicalizeOutput?: (response: Response) => Uint8Array;
  canonicalizeError?: (error: unknown) => Uint8Array;
  isDenied?: (request: Request) => boolean | Promise<boolean>;
  deniedResponse?: (request: Request) => Response | Promise<Response>;
};

export type SelloMcpToolInvocation<Args, Context = unknown> = {
  name: string;
  arguments: Args;
  context: Context;
};

export type SelloMcpToolHandler<Args, Response, Context = unknown> = (
  args: Args,
  context: Context,
) => Response | Promise<Response>;

export type SelloMcpToolOptions<Args, Response, Context = unknown> = {
  actionType?: string;
  authorizationToken?: SelloValueOrGetter<
    SelloMcpToolInvocation<Args, Context>,
    string | Uint8Array
  >;
  canonicalizeInput?: (
    invocation: SelloMcpToolInvocation<Args, Context>,
  ) => Uint8Array;
  canonicalizeOutput?: (response: Response) => Uint8Array;
  canonicalizeError?: (error: unknown) => Uint8Array;
  isDenied?: (
    invocation: SelloMcpToolInvocation<Args, Context>,
  ) => boolean | Promise<boolean>;
  deniedResponse?: (
    invocation: SelloMcpToolInvocation<Args, Context>,
  ) => Response | Promise<Response>;
};

export type SelloServiceConfig = {
  service?: string;
  serviceKey?: ServiceKeyInput;
  serviceKid?: KeyInput;
  tokenIssuer?: KeyInput | { publicKey?: KeyInput; jwksUrl?: string };
  tokenIssuerPublicKey?: KeyInput;
  tokenIssuerJwks?: string;
  log?: SdkSubmissionLog;
  fallbackSelloLogs?: readonly string[];
  submit?: SubmitMode | SelloSubmitOptions;
  now?: () => string;
  onReceipt?: (event: SelloReceiptEvent) => void;
  onSubmitError?: (error: unknown) => void;
  onDrop?: (event: DropEvent) => void;
};

export type SelloServiceInput = string | SelloServiceConfig;

export type SelloReceipts = {
  tool<Request, Response>(
    actionType: string,
    handler: SelloToolHandler<Request, Response>,
    options?: SelloToolOptions<Request, Response>,
  ): SelloToolHandler<Request, Response>;
  mcpTool<Args, Response, Context = unknown>(
    name: string,
    handler: SelloMcpToolHandler<Args, Response, Context>,
    options?: SelloMcpToolOptions<Args, Response, Context>,
  ): SelloMcpToolHandler<Args, Response, Context>;
  flush(): Promise<void>;
};

type Environment = Record<string, string | undefined>;

type TokenIssuerConfig =
  | { type: "public-key"; publicKey: Uint8Array }
  | { type: "jwks"; jwksUrl: string; publicKey?: Uint8Array };

type ResolvedServiceConfig = {
  service: string;
  serviceKid: Uint8Array;
  servicePrivateKey: Uint8Array;
  tokenIssuer: TokenIssuerConfig;
  log: SdkSubmissionLog;
  fallbackSelloLogs?: readonly string[];
  now: () => string;
  onReceipt?: (event: SelloReceiptEvent) => void;
};

type DeferredConfig =
  | { type: "resolved"; config: ResolvedServiceConfig }
  | {
      type: "hosted";
      secretKey: string;
      configUrl: string;
      now: () => string;
      onReceipt?: (event: SelloReceiptEvent) => void;
    };

export function createSelloService(input?: SelloServiceInput): SelloReceipts {
  const deferred = resolveDeferredConfig(input, process.env);
  const publisherOptions = resolvePublisherOptions(input, process.env);
  let loaded: Promise<ResolvedServiceConfig> | undefined;
  let publisher: BackgroundReceiptPublisher | undefined;

  async function config(): Promise<ResolvedServiceConfig> {
    loaded ??= loadDeferredConfig(deferred);
    return await loaded;
  }

  function publisherFor(resolved: ResolvedServiceConfig): BackgroundReceiptPublisher {
    publisher ??= new BackgroundReceiptPublisher({
      log: resolved.log,
      ...publisherOptions,
    });
    return publisher;
  }

  function wrapTool<Request, Response>(
    actionType: string,
    handler: SelloToolHandler<Request, Response>,
    options: SelloToolOptions<Request, Response> = {},
  ): SelloToolHandler<Request, Response> {
    if (typeof actionType !== "string" || actionType.length === 0) {
      throw new TypeError("Sello action type must be a non-empty string");
    }

    return async (request: Request): Promise<Response> => {
      const resolved = await config();
      const authorizationToken = resolveValue(
        options.authorizationToken ?? defaultAuthorizationToken,
        request,
      );
      const tokenIssuerPublicKey = await resolveTokenIssuerPublicKey(resolved.tokenIssuer);
      const verifiedToken = verifySelloJwsToken({
        authorizationToken,
        issuerPublicKey: tokenIssuerPublicKey,
      });
      const selloLogs = selectSelloLogs(
        verifiedToken.selloLogs,
        resolved.fallbackSelloLogs,
        resolved.log.logUrl,
      );
      const base = {
        authorizationTokenBytes: verifiedToken.authorizationTokenBytes,
        ownerHpkePublicKey: verifiedToken.ownerHpkePublicKey,
        selloLogs,
        serviceKid: resolved.serviceKid,
        servicePrivateKey: resolved.servicePrivateKey,
        serviceIdentifier: resolved.service,
        logUrl: resolved.log.logUrl,
        actionType,
        actionInputBytes: (options.canonicalizeInput ?? canonicalJsonBytes)(request),
        timestamp: resolved.now(),
      };

      if (options.isDenied && (await options.isDenied(request))) {
        const response = options.deniedResponse
          ? await options.deniedResponse(request)
          : undefined;
        const receipt = emitReceipt({
          ...base,
          actionOutputBytes: new Uint8Array(),
          resultStatus: "denied",
        });
        resolved.onReceipt?.({ resultStatus: "denied", receipt, response });
        await submit(resolved, receipt, base.timestamp);

        if (options.deniedResponse) {
          return response as Response;
        }

        throw new SelloDeniedError(receipt);
      }

      let response: Response;
      try {
        response = await handler(request);
      } catch (error) {
        const receipt = emitReceipt({
          ...base,
          actionOutputBytes: (options.canonicalizeError ?? canonicalErrorBytes)(
            error,
          ),
          resultStatus: "error",
        });
        resolved.onReceipt?.({ resultStatus: "error", receipt, error });
        await submit(resolved, receipt, base.timestamp);
        throw error;
      }

      const receipt = emitReceipt({
        ...base,
        actionOutputBytes: (options.canonicalizeOutput ?? canonicalJsonBytes)(
          response,
        ),
        resultStatus: "success",
      });
      resolved.onReceipt?.({ resultStatus: "success", receipt, response });
      await submit(resolved, receipt, base.timestamp);
      return response;
    };
  }

  return {
    tool<Request, Response>(
      actionType: string,
      handler: SelloToolHandler<Request, Response>,
      options: SelloToolOptions<Request, Response> = {},
    ): SelloToolHandler<Request, Response> {
      return wrapTool(actionType, handler, options);
    },
    mcpTool<Args, Response, Context = unknown>(
      name: string,
      handler: SelloMcpToolHandler<Args, Response, Context>,
      options: SelloMcpToolOptions<Args, Response, Context> = {},
    ): SelloMcpToolHandler<Args, Response, Context> {
      if (typeof name !== "string" || name.length === 0) {
        throw new TypeError("MCP tool name must be a non-empty string");
      }

      const wrapped = wrapTool<SelloMcpToolInvocation<Args, Context>, Response>(
        options.actionType ?? `mcp.tools/call.${name}`,
        async (invocation) => handler(invocation.arguments, invocation.context),
        {
          authorizationToken:
            options.authorizationToken ?? defaultMcpAuthorizationToken,
          canonicalizeInput:
            options.canonicalizeInput ?? defaultMcpCanonicalizeInput,
          canonicalizeOutput: options.canonicalizeOutput,
          canonicalizeError: options.canonicalizeError,
          isDenied: options.isDenied,
          deniedResponse: options.deniedResponse,
        },
      );

      return async (args: Args, context: Context): Promise<Response> => {
        return await wrapped({ name, arguments: args, context });
      };
    },
    async flush(): Promise<void> {
      await publisher?.flush();
    },
  };

  async function submit(
    resolved: ResolvedServiceConfig,
    receipt: BuiltReceipt,
    integratedTime: string,
  ): Promise<void> {
    const currentPublisher = publisherFor(resolved);
    if (currentPublisher.mode === "await") {
      await currentPublisher.publish({
        envelope: receipt.envelope,
        integratedTime,
      });
      return;
    }

    currentPublisher.publishBackground({
      envelope: receipt.envelope,
      integratedTime,
    });
  }
}

export class SelloDeniedError extends Error {
  readonly receipt: BuiltReceipt;

  constructor(receipt: BuiltReceipt) {
    super("Sello request denied");
    this.name = "SelloDeniedError";
    this.receipt = receipt;
  }
}

function emitReceipt(input: Parameters<typeof buildReceipt>[0]): BuiltReceipt {
  return buildReceipt(input);
}

function resolveDeferredConfig(
  input: SelloServiceInput | undefined,
  env: Environment,
): DeferredConfig {
  const objectInput =
    typeof input === "object" && input !== null ? input : {};
  const serviceOverride = typeof input === "string" ? input : objectInput.service;
  const hostedSecret = env.SELLO_SECRET_KEY;

  if (
    hostedSecret &&
    objectInput.serviceKey === undefined &&
    env.SELLO_SERVICE_KEY === undefined
  ) {
    return {
      type: "hosted",
      secretKey: hostedSecret,
      configUrl:
        env.SELLO_HOSTED_CONFIG_URL ?? "https://sello.build/api/sdk/config",
      now: objectInput.now ?? nowUtcSeconds,
      onReceipt: objectInput.onReceipt,
    };
  }

  return {
    type: "resolved",
    config: resolveSelfHostedConfig(objectInput, serviceOverride, env),
  };
}

function resolvePublisherOptions(
  input: SelloServiceInput | undefined,
  env: Environment,
): Omit<ConstructorParameters<typeof BackgroundReceiptPublisher>[0], "log"> {
  const objectInput =
    typeof input === "object" && input !== null ? input : {};

  return {
    submit: objectInput.submit ?? envSubmitMode(env),
    onSubmitError: objectInput.onSubmitError,
    onDrop: objectInput.onDrop,
  };
}

function resolveSelfHostedConfig(
  input: SelloServiceConfig,
  serviceOverride: string | undefined,
  env: Environment,
): ResolvedServiceConfig {
  const service = serviceOverride ?? env.SELLO_SERVICE_ID;
  if (!service) {
    throw new TypeError(
      "Sello setup missing SELLO_SERVICE_ID. Set SELLO_SERVICE_ID or call sello.service(\"service-id\").",
    );
  }

  const serviceKey = normalizeServiceKey(
    input.serviceKey ?? env.SELLO_SERVICE_KEY,
    input.serviceKid ?? env.SELLO_SERVICE_KID,
  );
  const log =
    input.log ??
    (env.SELLO_LOG_URL
      ? http(env.SELLO_LOG_URL, { endpoint: env.SELLO_LOG_ENDPOINT })
      : undefined);

  if (!log) {
    throw new TypeError(
      "Sello setup missing SELLO_LOG_URL. Set SELLO_LOG_URL or pass log explicitly.",
    );
  }

  return {
    service,
    serviceKid: serviceKey.kid,
    servicePrivateKey: serviceKey.privateKey,
    tokenIssuer: normalizeTokenIssuer(input, env),
    log,
    fallbackSelloLogs: input.fallbackSelloLogs ?? [log.logUrl],
    now: input.now ?? nowUtcSeconds,
    onReceipt: input.onReceipt,
  };
}

async function loadDeferredConfig(
  deferred: DeferredConfig,
): Promise<ResolvedServiceConfig> {
  if (deferred.type === "resolved") {
    return deferred.config;
  }

  const response = await fetch(deferred.configUrl, {
    method: "GET",
    headers: {
      authorization: `Bearer ${deferred.secretKey}`,
    },
  });

  if (!response.ok) {
    throw new TypeError(`sello.build config fetch failed with HTTP ${response.status}`);
  }

  const config = await response.json();
  if (!isRecord(config)) {
    throw new TypeError("sello.build config response must be an object");
  }

  const service = readString(config.service, "hosted service");
  const logUrl = readString(config.logUrl, "hosted logUrl");
  const hostedServiceKey = normalizeServiceKey(
    readString(config.serviceKey, "hosted serviceKey"),
  );

  return {
    service,
    serviceKid: hostedServiceKey.kid,
    servicePrivateKey: hostedServiceKey.privateKey,
    tokenIssuer: {
      type: "public-key",
      publicKey: normalizeEd25519PublicKey(
        readString(config.tokenIssuerPublicKey, "hosted tokenIssuerPublicKey"),
        "hosted tokenIssuerPublicKey",
      ),
    },
    log: http(logUrl, {
      endpoint:
        typeof config.logEndpoint === "string" ? config.logEndpoint : undefined,
    }),
    fallbackSelloLogs: [http(logUrl).logUrl],
    now: deferred.now,
    onReceipt: deferred.onReceipt,
  };
}

function normalizeTokenIssuer(
  input: SelloServiceConfig,
  env: Environment,
): TokenIssuerConfig {
  const tokenIssuer = input.tokenIssuer;

  if (tokenIssuer instanceof Uint8Array || typeof tokenIssuer === "string") {
    return {
      type: "public-key",
      publicKey: normalizeEd25519PublicKey(tokenIssuer, "tokenIssuer"),
    };
  }

  if (tokenIssuer && typeof tokenIssuer === "object") {
    if (tokenIssuer.publicKey) {
      return {
        type: "public-key",
        publicKey: normalizeEd25519PublicKey(
          tokenIssuer.publicKey,
          "tokenIssuer.publicKey",
        ),
      };
    }

    if (tokenIssuer.jwksUrl) {
      return { type: "jwks", jwksUrl: tokenIssuer.jwksUrl };
    }
  }

  if (input.tokenIssuerPublicKey || env.SELLO_TOKEN_ISSUER_PUBLIC_KEY) {
    return {
      type: "public-key",
      publicKey: normalizeEd25519PublicKey(
        input.tokenIssuerPublicKey ?? env.SELLO_TOKEN_ISSUER_PUBLIC_KEY as string,
        "SELLO_TOKEN_ISSUER_PUBLIC_KEY",
      ),
    };
  }

  const jwksUrl = input.tokenIssuerJwks ?? env.SELLO_TOKEN_ISSUER_JWKS;
  if (jwksUrl) {
    return { type: "jwks", jwksUrl };
  }

  throw new TypeError(
    "Sello setup missing token issuer. Set SELLO_TOKEN_ISSUER_PUBLIC_KEY, SELLO_TOKEN_ISSUER_JWKS, or pass tokenIssuer.",
  );
}

async function resolveTokenIssuerPublicKey(
  issuer: TokenIssuerConfig,
): Promise<Uint8Array> {
  if (issuer.type === "public-key") {
    return issuer.publicKey;
  }

  issuer.publicKey ??= await fetchEd25519JwksKey(issuer.jwksUrl);
  return issuer.publicKey;
}

async function fetchEd25519JwksKey(jwksUrl: string): Promise<Uint8Array> {
  const response = await fetch(jwksUrl);
  if (!response.ok) {
    throw new TypeError(`token issuer JWKS fetch failed with HTTP ${response.status}`);
  }

  const jwks = await response.json();
  if (!isRecord(jwks) || !Array.isArray(jwks.keys)) {
    throw new TypeError("token issuer JWKS must contain keys");
  }

  const key = jwks.keys.find(
    (candidate) =>
      isRecord(candidate) &&
      candidate.kty === "OKP" &&
      candidate.crv === "Ed25519" &&
      typeof candidate.x === "string",
  );

  if (!key || !isRecord(key) || typeof key.x !== "string") {
    throw new TypeError("token issuer JWKS must contain an Ed25519 OKP key");
  }

  return normalizeEd25519PublicKey(decodeBase64url(key.x, "JWKS x"), "JWKS x");
}

function selectSelloLogs(
  tokenLogs: readonly string[] | undefined,
  fallbackLogs: readonly string[] | undefined,
  logUrl: CanonicalLogUrl,
): readonly string[] {
  const selloLogs = tokenLogs ?? fallbackLogs ?? [];
  if (selloLogs.length === 0) {
    throw new TypeError("Sello token did not provide owner-trusted logs");
  }

  for (const entry of selloLogs) {
    assertCanonicalLogUrl(entry, "sello_logs entry");
    if (logUrlsEqual(entry, logUrl)) {
      return selloLogs;
    }
  }

  throw new TypeError("Sello log must be listed in the token's sello_logs");
}

function defaultAuthorizationToken(request: unknown): string | Uint8Array {
  const token = authorizationTokenFromUnknown(request);
  if (token) {
    return token;
  }

  throw new TypeError(
    "Sello authorization token not found. Pass authorizationToken or include request.authorizationToken.",
  );
}

function defaultMcpAuthorizationToken(
  invocation: SelloMcpToolInvocation<unknown, unknown>,
): string | Uint8Array {
  const fromContext = authorizationTokenFromUnknown(invocation.context);
  if (fromContext) {
    return fromContext;
  }

  throw new TypeError(
    "Sello MCP authorization token not found. Pass authorizationToken or include an Authorization header in the MCP context.",
  );
}

function defaultMcpCanonicalizeInput(
  invocation: SelloMcpToolInvocation<unknown, unknown>,
): Uint8Array {
  return canonicalJsonBytes({
    method: "tools/call",
    params: {
      name: invocation.name,
      arguments: invocation.arguments,
    },
  });
}

function authorizationTokenFromUnknown(value: unknown): string | Uint8Array | undefined {
  if (!isRecord(value)) {
    return undefined;
  }

  const direct = value.authorizationToken ?? value.authorization;
  if (typeof direct === "string" || direct instanceof Uint8Array) {
    return stripBearer(direct);
  }

  const headers = value.headers;
  const fromHeaders = authorizationTokenFromHeaders(headers);
  if (fromHeaders) {
    return fromHeaders;
  }

  const requestInfo = value.requestInfo;
  if (isRecord(requestInfo)) {
    const fromRequestInfo = authorizationTokenFromHeaders(requestInfo.headers);
    if (fromRequestInfo) {
      return fromRequestInfo;
    }
  }

  const request = value.request;
  if (isRecord(request)) {
    const fromRequest = authorizationTokenFromHeaders(request.headers);
    if (fromRequest) {
      return fromRequest;
    }
  }

  const authInfo = value.authInfo;
  if (isRecord(authInfo)) {
    const token = authInfo.token ?? authInfo.accessToken ?? authInfo.access_token;
    if (typeof token === "string" || token instanceof Uint8Array) {
      return stripBearer(token);
    }
  }

  return undefined;
}

function authorizationTokenFromHeaders(
  headers: unknown,
): string | Uint8Array | undefined {
  if (isHeadersLike(headers)) {
    const header = headers.get("authorization") ?? headers.get("Authorization");
    if (typeof header === "string") {
      return stripBearer(header);
    }
  }

  if (!isRecord(headers)) {
    return undefined;
  }

  const header = headers.authorization ?? headers.Authorization;
  if (typeof header === "string" || header instanceof Uint8Array) {
    return stripBearer(header);
  }

  return undefined;
}

function isHeadersLike(
  value: unknown,
): value is { get(name: string): string | null } {
  return (
    typeof value === "object" &&
    value !== null &&
    typeof (value as { get?: unknown }).get === "function"
  );
}

function stripBearer(value: string | Uint8Array): string | Uint8Array {
  if (typeof value !== "string") {
    return value;
  }

  return value.startsWith("Bearer ") ? value.slice("Bearer ".length) : value;
}

function canonicalErrorBytes(error: unknown): Uint8Array {
  if (error instanceof Error) {
    return canonicalJsonBytes({
      name: error.name,
      message: error.message,
    });
  }

  return canonicalJsonBytes({ error: String(error) });
}

function resolveValue<Request, Value>(
  value: SelloValueOrGetter<Request, Value>,
  request: Request,
): Value {
  if (typeof value === "function") {
    return (value as (request: Request) => Value)(request);
  }

  return value;
}

function envSubmitMode(env: Environment): SubmitMode | SelloSubmitOptions | undefined {
  const mode = env.SELLO_SUBMIT_MODE;
  if (mode === undefined) {
    return undefined;
  }

  if (mode !== "background" && mode !== "await") {
    throw new TypeError("SELLO_SUBMIT_MODE must be background or await");
  }

  return mode;
}

function nowUtcSeconds(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function readString(value: unknown, name: string): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new TypeError(`${name} must be a non-empty string`);
  }

  return value;
}
